### ISSUE TITLE
Is pytorch SGD optimizer apply weight decay to bias parameters with default settings?

### ISSUE THREAD

**Issue by dianyancao** [pytorch/pytorch#2639]
Generally,bias does not cause overfitting.I look into the function 'SGD.step' in 'torch\optim\sgd.py',but can't confirm it separates weight and bias parameters.

**Comment by CDLuminate** [comment]
The optimizer will apply weight decay to the biases.

The optimizer is instantiated with a set of parameters, and the set of parameters contains biases.
```
In [1]: import torch as th

In [2]: th.nn.Linear(10, 10)
Out[2]: Linear (10 -> 10)

In [6]: for x in th.nn.Linear(10, 10).parameters(): print(x.size())
torch.Size([10, 10])
torch.Size([10])
```

optimizer.py
```
 28         self.param_groups = list(params)
```

sgd.py
```
 74         for group in self.param_groups:                                         
 75             weight_decay = group['weight_decay']                                
 76             momentum = group['momentum']                                        
 77             dampening = group['dampening']                                      
 78             nesterov = group['nesterov']                                        
 79                                                                                 
 80             for p in group['params']:                                           
 81                 if p.grad is None:                                              
 82                     continue                                                    
 83                 d_p = p.grad.data                                               
 84                 if weight_decay != 0:                                           
 85                     d_p.add_(weight_decay, p.data)   
```

The biases will be decayed when the optimizer traversed to it.

**Comment by dianyancao** [comment]
Is it good to apply weight decay to biases? @CDLuminate

**Comment by CDLuminate** [comment]
Hard to say. There is no best method of regularization. Since the intention of weight decay (usually L2 regularization) is to decrease the L2 norm of weights to shrink the hypothesis space, I'd suggest that the biases be decayed.

**Comment by dianyancao** [comment]
OK,got it,Thank you!

**Comment by nnop** [comment]
PRML says:
> Note that often the coefficient w0 is omitted from the regularizer because its inclusion causes the results to depend on the choice of origin for the target variable (Hastie et al., 2001), or it may be included but with its own regularization coefficient (we shall discuss this topic in more detail in Section 5.5.1).
